
<h1 align="center"> MIntO </h1>
<h1 align="center"> A Modular and Scalable Pipeline for Microbiome Metagenomics and Metatranscriptomics Data Integration </h1>

MIntO (Microbiome Integrated Meta-Omics), is a **highly versatile pipeline** that integrates metagenomic and metatranscriptomic data in a scalable way.
The distinctive feature of this pipeline is the computation of gene expression profile by taking into account the community turnover and gene expression variations as the underlying process that shapes the community transcript levels along the time and between conditions.
The integrated pipeline will be relevant to provide unique biochemical insights into the microbial ecology by linking functions to retrieved genomes and to examine gene expression variations. 

MIntO can be downloaded here:  https://github.com/arumugamlab/MIntO

<p align="center"><img src="images/MIntO_blu_Microbiome_integrated_meta_omics.png" height="200" /></p>


## What should I know about MIntO?

### You do not need to be a programmer to use MIntO

MIntO aims to reduce the barrier of entry for metagenomic/metatranscriptomic data analysis. Experience with linux/unix command line is required, but that's just about it.
MIntO is designed to work out of the box, as long as you provide the right information in the configuration files - this is explained in the tutorials.
We provide all the key steps and necessary explanations in the [wiki section](https://github.com/arumugamlab/MIntO/wiki/). 

### If you are a programmer, you can also use MIntO

If you are experienced with bioinformatic pipelines and have used Snakemake, then you could tweak MIntO for your own purposes. You can even add new functionality. You are welcome to contribute to the repository if you are interested.

### How do I get started?

We suggest to read the **[MIntO in a nutshell](https://github.com/arumugamlab/MIntO/wiki/MIntO-in-a-nutshell)** section first, just to have an idea on how MIntO works. 
Then move to **[Installation&Dependencies](https://github.com/arumugamlab/MIntO/wiki/Installation&Dependencies)** and then you are set to discover each step! 
Also you can have a look to the **[tutorial](https://github.com/arumugamlab/MIntO/wiki/MIntO-tutorial)**.


## Publication

If you have used MIntO during your analysis, please cite:

 **MIntO: a Modular and Scalable Pipeline for Microbiome Meta-omics Data Integration**  
Carmen Saenz, Eleonora Nigro, Vithiagaran Gunalan, & Manimozhiyan Arumugam  
_Frontiers in Bioinformatics_ (2022). doi: [10.3389/fbinf.2022.846922](https://doi.org/10.3389/fbinf.2022.846922). 


## Contacts

Feel free to get in touch with the authors if you have any questions. You can post issues in this repository, or write to the corresponding author from the article link above.

=======

